### PR TITLE
[DPP-142] Explicitly deflate/inflate data outside of the index

### PR DESCRIPTION
**Pull request by mzagorski-da**
Compressing data manually before persisting into database.

Changes brings ~13.3% tps throughput improvement in LR.

Comparing to original  #8039 , visible improvement was brought by adding parallelism and buffering to execute update stream in ExecuteUpdate.scala:

```
   private[indexer] val flow: ExecuteUpdateFlow =
    Flow[OffsetUpdate]
      .mapAsync(2)(prepareUpdate)
      .buffer(16, OverflowStrategy.backpressure)
``` 

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTableH2Database.scala [hunk 1]
```diff
@@ -77,74 +78,119 @@ object EventsTableH2Database extends EventsTable {
...
+    //        "node_index" -> nodeId.index,
+    //        "flat_event_witnesses" -> Party.Array(stakeholders.toSeq: _*),
+    //        "tree_event_witnesses" -> Party.Array(disclosure.toSeq: _*),
+    //      )
+    //    val nonShared =
+    //      node match {
+    //        case event: Create => create(event, createArgument.get, createKeyValue)
+    //        case event: Exercise => exercise(event, transactionId, exerciseArgument.get, exerciseResult)
+    //        case _ => throw new UnexpectedNodeException(nodeId, transactionId)
+    //=======
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTableH2Database.scala:118]: Hmm

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala [hunk 2]
```diff
@@ -372,7 +373,8 @@ class PipelinedExecuteUpdate(
 
   private[indexer] val flow: ExecuteUpdateFlow =
     Flow[OffsetUpdate]
-      .mapAsync(1)(prepareUpdate)
+      .mapAsync(2)(prepareUpdate)
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala:376]: I think the parallelism of this stage should be configurable.

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres/V42__explicit_compression.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres/V42__explicit_compression.sql:1]: Can you update the copyright?

**Review by stefanobaghino-da — COMMENTED** [review]
Looks ok. Can you summarize the improvements over https://github.com/digital-asset/daml/pull/8039 that gave this PR a boost? I can't tell right away.

**Comment by mzagorski-da** [comment]
> Looks ok. Can you summarize the improvements over #8039 that gave this PR a boost? I can't tell right away.

I added info to pr description:

```
Comparing to original #8039 , visible improvement was brought by adding parallelism and buffering to execute update stream in ExecuteUpdate.scala:

   private[indexer] val flow: ExecuteUpdateFlow =
    Flow[OffsetUpdate]
      .mapAsync(2)(prepareUpdate)
      .buffer(16, OverflowStrategy.backpressure)
```

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres/V42__explicit_compression.sql (hunk 3 above)
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres/V42__explicit_compression.sql]: fixed

File: ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala (hunk 2 above)
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala]: exposed in `com.daml.platform.indexer.IndexerConfig`

**Review by tudor-da — APPROVED** [review]

**Comment by stefanobaghino-da** [comment]
> > Looks ok. Can you summarize the improvements over #8039 that gave this PR a boost? I can't tell right away.
> 
> I added info to pr description:
> 
> ```
> Comparing to original #8039 , visible improvement was brought by adding parallelism and buffering to execute update stream in ExecuteUpdate.scala:
> 
>    private[indexer] val flow: ExecuteUpdateFlow =
>     Flow[OffsetUpdate]
>       .mapAsync(2)(prepareUpdate)
>       .buffer(16, OverflowStrategy.backpressure)
> ```

Thanks for the additional info. Did you check whether these advantages appear only if coupled with the introduction of out-of-database compression?

**Comment by mzagorski-da** [comment]
@stefanobaghino-da 

> > > Looks ok. Can you summarize the improvements over #8039 that gave this PR a boost? I can't tell right away.
> > 
> > 
> > I added info to pr description:
> > ```
> > Comparing to original #8039 , visible improvement was brought by adding parallelism and buffering to execute update stream in ExecuteUpdate.scala:
> > 
> >    private[indexer] val flow: ExecuteUpdateFlow =
> >     Flow[OffsetUpdate]
> >       .mapAsync(2)(prepareUpdate)
> >       .buffer(16, OverflowStrategy.backpressure)
> > ```
> 
> Thanks for the additional info. Did you check whether these advantages appear only if coupled with the introduction of out-of-database compression?


In fact we didn't but improvement was visible in other metrics. I run it today to confirm. It looks like buffering brings some improvement, especially at the beginning of the run, but with compression, improvement is even bigger and permanent. Results below.

Run with this pr:

https://lr.da-ext.net/grafana/d/HfZvOTCZz/daml-on-in-memory-pkv?orgId=1&from=1611927389980&to=1611927611371&var-benchmark_id=143

Run with parallel prepare update + buffer only:

https://lr.da-ext.net/grafana/d/HfZvOTCZz/daml-on-in-memory-pkv?orgId=1&from=1612188970355&to=1612189173998&var-benchmark_id=159

**Comment by stefanobaghino-da** [comment]
@mzagorski-da Thanks for the detailed explanation!

File: ledger/participant-integration-api/src/main/scala/platform/store/serialization/Compression.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  sealed abstract class Algorithm(val id: Option[Int]) {
+    final def compress(
+        uncompressed: Array[Byte],
+        metrics: CompressionMetrics.Field,
+    ): Array[Byte] = {
+      val output = new ByteArrayOutputStream(uncompressed.length)
+      val gzip = compress(output)
+      try {
+        gzip.write(uncompressed)
+      } finally {
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/serialization/Compression.scala:22]: @mzagorski-da : this swallows an exception and then just pretends everything was allright doesn't it?

That sounds like a bad problem in waiting. Can you clarify how this code is supposed to work in that case?

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/serialization/Compression.scala:22]: @meiersi-da I'm confused, where is the exception swallowed?

Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/serialization/Compression.scala:22]: o.O -- I'm outing myself as a Java noob here :/ -- please ignore my comment. `finally` obviously lets the exception bubble up.

:see_no_evil:
